src/prune.ts

```diff
@@ -0,0 +1,247 @@
+              typingsPath: join(typingsPath, MAIN_TYPINGS_DIR),
+              emitter
+            }),
+          ]).then(() => { }) // convert to Promise<void>
+        })
+    })
+}
+
+interface TypeDependencies {
+  ambientDependencies: string[]
+  externalDependencies: string[]
+}
+
+function readMasterDependencies(projectDirectory: string, includeDev: boolean): Promise<TypeDependencies> {
+  return readConfig(join(projectDirectory, CONFIG_FILE))
+    .then((config) => {
+      const ambientDependencies = config.ambientDependencies != null
+        ? Object.keys(config.ambientDependencies)
+        : []
+
+      const externalDependencies = config.dependencies != null
+        ? Object.keys(config.dependencies)
+        : []
+
+      const dependencies: TypeDependencies = {
+        ambientDependencies,
+        externalDependencies
+      }
+
+      if (includeDev) {
+        const ambientDevDependencies = config.ambientDevDependencies != null
+          ? Object.keys(config.ambientDevDependencies)
+          : []
+
+        const externalDevDependencies = config.devDependencies != null
+          ? Object.keys(config.devDependencies)
+          : []
+
+        dependencies.ambientDependencies = dependencies.ambientDependencies.concat(ambientDevDependencies)
+        dependencies.externalDependencies = dependencies.externalDependencies.concat(externalDevDependencies)
+      }
+
+      return dependencies
+    })
+}
+
+interface PruneReferencesOptions {
+  masterDependencies: TypeDependencies
+  project: string
+  dtsPath: string
+  emitter: Emitter
+}
+
+function pruneExtraneousReferences(options: PruneReferencesOptions) {
+  const { masterDependencies, dtsPath, emitter } = options
+
+  return readFile(dtsPath, 'utf8')
+    .then((config: string) => {
+      const lines = config.split('\n')
+      // Split into extraneous and non-extraneous
+      const partitionedLines = partition(lines, (line: string) => isExtraneousReference(masterDependencies, line))
+      const [extraneousTypings, usedTypings] = partitionedLines
+
+      // Write out references that are being pruned
+      extraneousTypings.forEach((line: string) => emitter.emit('pruneReference', dtsPath, line, getReferenceType(line)))
+
+      // Join remaining lines and write back to config
+      const prunedConfig = usedTypings.join('\n')
+      return writeFile(dtsPath, prunedConfig)
+    })
+}
+
+function isExtraneousReference(masterDependencies: TypeDependencies, reference: string): boolean {
+  const pathParts = getReferencePathParts(reference)
+  if (pathParts) {
+    const [ , ambientOrExternal, actualDependency] = pathParts
+
+    const matchingMasterDeps = (ambientOrExternal == AMBIENT_DEFINITIONS_DIR)
+      ? masterDependencies.ambientDependencies
+      : masterDependencies.externalDependencies
+
+    return !includes(matchingMasterDeps, actualDependency)
+  }
+
+  return false
+}
+
+function includes<T>(array: T[], item: T): boolean {
```

Comment:
I'd even just go with checking object keys. Simpler and faster. Either `typeof config.dependencies[x] === 'string'` or `hasOwnProperty` check. But that's a personal preference to work with maps usually rather than converting to an array.